### ISSUE TITLE
MA0082 - NaN should not be used in comparisons is handled by CA2242

### ISSUE THREAD

**Issue by Meir017** [meziantou/Meziantou.Analyzer#856]
- **Rule**: [MA0082](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/Rules/MA0082.md) - NaN should not be used in comparisons
- **Built-in Equivalent**: [CA2242](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2242) - Test for NaN correctly

## Analysis

**MA0082** and **CA2242** serve the same purpose: ensuring that `NaN` (Not a Number) values are tested correctly using appropriate methods rather than direct equality comparisons, which always return `false` for NaN.

### MA0082 Coverage
- Detects direct comparisons with NaN values
- Promotes proper NaN testing patterns
- Prevents logical errors in floating-point comparisons

### CA2242 Coverage  
- Detects incorrect NaN testing patterns
- Official Microsoft rule for proper floating-point handling
- Ensures correct NaN comparison methods are used
- Part of the .NET reliability guidelines

## Code Examples

**Problematic code detected by both rules:**
```csharp
// Both MA0082 and CA2242 will flag these
public class MathHelper
{
    public bool IsValid(double value)
    {
        // Wrong - this always returns false!
        if (value == double.NaN) return false;
        
        // Wrong - this always returns true!
        if (value != double.NaN) return true;
        
        return true;
    }
    
    public bool CheckFloat(float value)
    {
        // Wrong - equality with NaN always false!
        return value == float.NaN;
    }
}
```

**Recommended fix:**
```csharp
// Both rules suggest using proper NaN testing methods
public class MathHelper
{
    public bool IsValid(double value)
    {
        // Correct - use IsNaN method
        if (double.IsNaN(value)) return false;
        
        // Correct - use IsNaN method
        if (!double.IsNaN(value)) return true;
        
        return true;
    }
    
    public bool CheckFloat(float value)
    {
        // Correct - use IsNaN method
        return !float.IsNaN(value);
    }
    
    public bool IsFinite(double value)
    {
        // Also consider other useful methods
        return double.IsFinite(value); // Not NaN, not infinite
    }
}
```

## IEEE 754 Floating-Point Behavior

NaN has special properties:
- **Any comparison with NaN returns `false`** (including `NaN == NaN`)
- **`NaN != NaN` returns `true`** (unique property)
- **Mathematical operations with NaN produce NaN**
- **Only `IsNaN()` reliably detects NaN values**

## Common NaN Sources
```csharp
double nan1 = 0.0 / 0.0;           // Division by zero
double nan2 = Math.Sqrt(-1.0);     // Square root of negative
double nan3 = Math.Log(-1.0);      // Logarithm of negative
double nan4 = double.PositiveInfinity - double.PositiveInfinity; // Infinity arithmetic
```

## Recommendation

**CA2242** is the official Microsoft analyzer rule that provides **equivalent functionality** to MA0082. Since both rules target the same floating-point correctness issue and CA2242 is the standard rule for NaN handling, **MA0082 should be deprecated** in favor of CA2242.

## Migration Path

1. Enable **CA2242** in your analyzer configuration:
   ```xml
   <WarningsAsErrors Include="CA2242" />
   ```
2. Disable **MA0082**:
   ```xml
   <WarningsNotAsErrors Include="MA0082" />
   ```
3. Update any rule suppressions from MA0082 to CA2242

## Related

- Parent issue: #835 - Consider deprecating rules that are now covered by built-in .NET analyzers

**Comment by Meir017** [comment]
## Analysis: Equivalent Rules - Recommend Deprecation ✅

Continuing final batch analysis - all equivalent rules:

- **MA0082 vs CA2242** - NaN comparison correctness
- **MA0095 vs CA1067** - Equals override consistency  
- **MA0102 vs IDE0044** - Readonly field optimization
- **MA0112 vs CA1860** - Collection performance (Count vs Any)
- **MA0131 vs CA2264** - ThrowIfNull parameter validation
- **MA0142 vs IDE0041** - Modern null pattern matching

### Recommendation: Deprecate All ✅

Official Microsoft rules provide equivalent functionality.

---
*Part of #835 deprecation initiative.*

**Comment by Meir017** [comment]
Thank you for bringing this to attention! After analyzing both rules, I can confirm that **CA2242** and **MA0082** are already documented as equivalent rules in the [comparison file](https://github.com/meziantou/Meziantou.Analyzer/blob/main/docs/comparison-with-other-analyzers.md#equivalent-rules) (line 36).

## Analysis Summary

Both rules detect incorrect NaN comparisons and suggest using `IsNaN()` methods:

**CA2242** (Microsoft): Covers `System.Single.NaN` and `System.Double.NaN`
**MA0082** (Meziantou): Covers `double.NaN`, `float.NaN`, and `Half.NaN`

## Key Finding

**MA0082 actually provides slightly broader coverage** than CA2242 because it explicitly supports `Half.NaN` comparisons in addition to the standard float/double types. This gives MA0082 an advantage over the built-in analyzer rule.

## Recommendation

Since MA0082 covers **equivalent or broader scope** compared to CA2242, and this relationship is already documented in the comparison file, I'm closing this issue. The Meziantou Analyzer rule should be retained as it provides value beyond the built-in .NET analyzer.

The comparison documentation correctly categorizes these as "equivalent" rules, acknowledging they serve the same fundamental purpose while MA0082 offers additional coverage.
